### DISCUSSION TITLE
Consistent update statements order

### DISCUSSION THREAD

**Discussion by shodo** [sqlalchemy/sqlalchemy#7553] [Usage Questions] [answered]
Hi, 
I'm wondering if SQLAlchemy is reliable in terms of generating update statements always in a specific order (it could be useful to mitigate deadlock without explicit row locking). 
Seems it's not cause if i run this code a couple of times:

``` python
mario = UserModel(name="mario")  
luigi = UserModel(name="luigi")
peach = UserModel(name="peach")

session.add(mario)
session.add(luigi)
session.add(peach)
session.flush()

mario.name = "updated mario"
luigi.name = "updated luigi"
peach.name = "updated peach"
session.flush()
```

In the log I obtain on a first run:

``` bash
INFO  [sqlalchemy.engine.Engine] INSERT INTO user (id, name) VALUES (%s, %s)
INFO  [sqlalchemy.engine.Engine] [generated in 0.00033s] (1, 'mario'), (2, 'luigi'), (3, 'peach') 
INFO  [sqlalchemy.engine.Engine] UPDATE user SET name=%s WHERE user.id = %s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00073s] (('updated luigi', 2), ('updated peach', 3), ('updated mario', 1))
```
and on a second run:

``` bash
INFO  [sqlalchemy.engine.Engine] INSERT INTO user (id, name) VALUES (%s, %s)
INFO  [sqlalchemy.engine.Engine] [generated in 0.00033s] (1, 'mario'), (2, 'luigi'), (3, 'peach') 
INFO  [sqlalchemy.engine.Engine] UPDATE user SET name=%s WHERE user.id = %s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00073s] (('updated luigi', 2), ('updated mario', 1), ('updated peach', 3))
```

So:
- it doesn't reflect "code update order": first Mario, second Luigi, third Peach
- on every run the order changes: (Luigi, Peach, Mario) and then (Luigi, Mario, Peach)

**Question 1:**
is there any way to ensure that a certain order is respected relying on the ORM api?

Things are different if I create new instances merging them in the session like this:

``` python
mario = UserModel(name="mario")  
luigi = UserModel(name="luigi")
peach = UserModel(name="peach")

session.add(mario)
session.add(luigi)
session.add(peach)
session.flush()

new_mario = UserModel(id=mario.id, name="new mario")  
new_luigi = UserModel(id=luigi.id, name="new luigi")
new_peach = UserModel(id=peach.id, name="new peach")

session.merge(new_mario)
session.merge(new_luigi)
session.merge(new_peach)
session.flush()
```

In this case the output in terms of generated log is always the same:
``` bash
INFO  [sqlalchemy.engine.Engine] INSERT INTO user (id, name) VALUES (%s, %s)
INFO  [sqlalchemy.engine.Engine] [generated in 0.00031s] (1, 'mario'), (2, 'luigi'), (3, 'peach') 
INFO  [sqlalchemy.engine.Engine] UPDATE user SET name=%s WHERE user.id = %s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00111s] ('new mario', 1)
INFO  [sqlalchemy.engine.Engine] UPDATE user SET name=%s WHERE user.id = %s
INFO  [sqlalchemy.engine.Engine] [generated in 0.001199s] ('new luigi', 2)
INFO  [sqlalchemy.engine.Engine] UPDATE user SET name=%s WHERE user.id = %s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00090s] ('new peach', 3)
```

In this case the generated UPDATE statements follow the same order of how user are merged into the session. 
So, I have this other questions:

**Question 2:**
Is this order guaranteed?

**Question 3:**
Why in the second case I have in the log multiple UPDATE statements instead of one UPDATE statement with the list of tuple containing the params for each user?

**Question 4:**
Is the session flushed every time I call merge? Is it something that should be avoided?

**Comment by zzzeek** [comment, marked as answer]
> Hi, I'm wondering if SQLAlchemy is reliable in terms of generating update statements always in a specific order (it could be useful to mitigate deadlock without explicit row locking). Seems it's not cause if i run this code a couple of times:

that's not right at all, and I cannot reproduce.   Please share SQLAlchemy version info as well as Python version in use.

 Objects that are within the same mapped class (and therefore table) are sorted by primary key and my [MCVE](http://stackoverflow.com/help/mcve) below on my end sorts them by PK and the order of operations is deterministic.   The sorting occurs [here](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/persistence.py#L372) and [here](https://github.com/sqlalchemy/sqlalchemy/blob/5681d4e4da8ee69d83e9c0103c171d413d4c183e/lib/sqlalchemy/orm/persistence.py#L1714).


> **Question 1:** is there any way to ensure that a certain order is respected relying on the ORM api?

UPDATE statements within a single flush for a particular mapped class hierarchy are always sorted by the primary key established on the objects.

> In this case the generated UPDATE statements follow the same order of how user are merged into the session. So, I have this other questions:
> 
> **Question 2:** Is this order guaranteed?

it's guaranteed in both cases, within the scope of a single flush() call.   if there are multiple flushes being triggered by end-user code, it UPDATEs what's ready to be updated so ordering between multiple rows in those separate flushes are driven by the end-user code.

> 
> **Question 3:** Why in the second case I have in the log multiple UPDATE statements instead of one UPDATE statement with the list of tuple containing the params for each user?

each merge() is likely being autoflushed.

> 
> **Question 4:** Is the session flushed every time I call merge? Is it something that should be avoided?

if autoflush is enabled and the merge() operation needs to emit SQL then yes.   merge() is usually something I avoid as it incurs a SELECT per row.  it's better to load all the objects you need up front, make modifications to them all at once, then commit.  if you need to control when flush occurs, use the [session.no_autoflush context manager](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=no_autoflush#sqlalchemy.orm.Session.no_autoflush).

MCVE for add() + update() being ordered:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()

class UserModel(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    name = Column(String)




e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

session = Session(e)

mario = UserModel(name="mario")
luigi = UserModel(name="luigi")
peach = UserModel(name="peach")

session.add(mario)
session.add(luigi)
session.add(peach)
session.flush()

mario.name = "updated mario"
luigi.name = "updated luigi"
peach.name = "updated peach"
session.flush()
```

output is every time, including when setting PYTHONHASHSEED to true:

```
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine INSERT INTO a (name) VALUES (?)
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine [generated in 0.00010s] ('mario',)
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine INSERT INTO a (name) VALUES (?)
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine [cached since 0.0003382s ago] ('luigi',)
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine INSERT INTO a (name) VALUES (?)
2022-01-09 14:08:08,712 INFO sqlalchemy.engine.Engine [cached since 0.000487s ago] ('peach',)
2022-01-09 14:08:08,714 INFO sqlalchemy.engine.Engine UPDATE a SET name=? WHERE a.id = ?
2022-01-09 14:08:08,714 INFO sqlalchemy.engine.Engine [generated in 0.00010s] (('updated mario', 1), ('updated luigi', 2), ('updated peach', 3))
```

**Reply by shodo** [reply]
Hi @zzzeek! 
Thanks for the fast reply. 
It was my mistake: I've adapted the example from a production code that uses UUID instead of integers for the primary key. This is why on each run the order was different!

Anyway, also running the right code and observing always the same order, I couldn't be sure that was _guaranteed_. Glad to know it's deterministic!

Regarding the "merge" I've asked cause we are using a sort of clean architecture in which we have domain entities that are not directly mapped to the one traced by the SQLAlchemy session. So we have every time to go back and forth from the SQLAlchemy models to our domain model, ending up with something similar to this:

```python

mario = DomainUser(name="mario")  
luigi = DomainUser(name="luigi")
peach = DomainUser(name="peach")

session.add(UserModel.from_domain(mario))
session.add(UserModel.from_domain(luigi))
session.add(UserModel.from_domain(peach))
session.flush()

mario.name = "updated mario"
luigi.name = "updated luigi"
peach.name = "updated peach"

sqla_mario = mario.to_sql_alchemy_model()
sqla_luigi = luigi.to_sql_alchemy_model()
sqla_peach = peach.to_sql_alchemy_model()
await session.merge(sqla_mario)
await session.merge(sqla_luigi)
await session.merge(sqla_peach)

session.flush()
```

Every time I made a change on the "domain" model I have to convert it into the "SQLAlchemy" model and reconcile it with the session. 
In another project, I've used SQLAlchemy imperative mapping to directly map the ORM on our domain entities, it worked like a charm, but I cannot use it anymore, cause of "internal company guidelines". 

Is there any way you can advise to avoid a single flush for each merge in case of multiple updates? If I simply rewrite the code as:

```python
  sqla_mario = mario.to_sql_alchemy_model()
  sqla_luigi = mario.to_sql_alchemy_model()
  sqla_peach = mario.to_sql_alchemy_model()
  
with session.no_autoflush:
  await session.merge(sqla_mario)
  await session.merge(sqla_luigi)
  await session.merge(sqla_peach)

  session.flush()
```

should work? Or it's better to do the merge work manually? I mean:
- make a statement with a WHERE user.id IN (mario.id, luigi.id, peach.id) to obtain SQLALchemy models
- update each model with the changes from the domain
- flush

Any other advice on implementing clean architecture with SQLAlchemy?

Thanks again A LOT for your time!

**Reply by zzzeek** [reply]
merge() is fine as far as SQL overhead if you know that the primary key being targeted it already present in the identity map.

so there's a pattern for this you can do like this:

```py
# this where you create the new data to merge, from your 
# JSON / XML / whatever
objects_to_merge = get_objects_to_merge()

# now load all those PKs into a list in memory.  they are now in the identity map
# keyed to pk.  make sure objects_from_the_db stays around in memory else
# the objects get garbage collected
objects_from_the_db = session.scalars(select(UserModel).where(UserModel.id.in_([obj.id for obj in objects_to_merge])).all()


for obj in objects_to_merge:
    new_obj = session.merge(obj)  # will give you an object that's in objects_to_merge
```

the merge() function is still merging the attributes but you've given it a hint ahead of time that loaded all those PKs first.  this is at least how merge can be "ok" if you have the pattern of receiving full objects from another source.

the creation of the objects themselves has performance overhead, so this pattern still involves creating each object twice.
